### ISSUE TITLE
Error when  moving exclusive image ownership between queue families with simultaneous image transition

### ISSUE THREAD

**Issue by ParticlePeter** [KhronosGroup/Vulkan-LoaderAndValidationLayers#865]
The issue is discussed in detail [here](https://forums.khronos.org/showthread.php/13153-Issue-moving-exclusive-image-ownership-between-queue-families-and-synchronization).
Summarizing the scenario:
1.) write into an `VK_SHARING_MODE_EXCLUSIVE` image from graphics queue family
2.) transition image to another queue family with compute capability for reading

[As per spec](https://www.khronos.org/registry/vulkan/specs/1.0-wsi_extensions/xhtml/vkspec.html#resources-sharing) two similar pipeline image barriers are required, one in the source queue for releasing the ownership and one in the destination queue for acquiring the ownership. The spec does not mention that source and destination layout must not differ, hence this might be an spec issue rather then a validation issue.
However, if the layouts do differ (in my case layout transition from `VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL` to `VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL`) validation reports this error (layout transition applied twice?):

`Cannot submit cmd buffer using image (0xa) [sub-resource: aspectMask 0x1 array layer 0, mip level 0], with layout VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL when first use is VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL.`

No error is reported when the layout transition happens elsewhere (render pass) and old and new layout are both and destination layout `VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL` in both of the pipeline barriers.

**Comment by lentinem** [comment]
I'm a bit confused. Is the layout transiton applied twice? If that is the case then the error is correct as you cannot transition an image from COLOR_ATTACHMENT if it's in layout SHADER_READ_ONLY. Otherwise it would help to have a repo case so we can track down what is happening. Thanks!

**Comment by ParticlePeter** [comment]
My example is inside a D abstraction layer and I doubt you want to read it. Let's take a look into the spec first and try to clarify before I create an example.
[Per Spec: Resource Sharing Mode](https://www.khronos.org/registry/vulkan/specs/1.0-wsi_extensions/xhtml/vkspec.html#resources-sharing), in particular this sentence:

> To acquire exclusive ownership, the application must execute the same buffer or image memory barrier on a queue from the destination queue family.

what exactly is meant by "the same" and how do image layout transitions fit into this concept?

1. Both barriers need to specify the same layout transition, otherwise they would not be "same". This might confuse drivers and/or layers due to the double transition.
2. None of the barriers is allowed to specify a layout transition, all of the four barrier layout members (2*(```oldLayout```+```newLayout```)) must match, implying that layout transitions are not allowed in this particular case. Barriers would be "the same", but this needs to be mentioned in the spec.  
3. We can _assume_ that the first (releasing) barrier is evaluated before the second (acquiring) barrier, hence only the ```newLayout``` of the fist barrier must match to the ```oldLayout``` of the second barrier. Barriers would not be "same", and based on the current spec this seems to be a big _assume_.

I tried 1. and 2., the former not working while the later does. 3. might work but should definitely be explained in the spec. Can you clarify which of the three cases is supposed to work and which not, please?

**Comment by krOoze** [comment]
I will try to reproduce it in a C++ example soon<sup>TM</sup>...

**Comment by mark-lunarg** [comment]
We interpret it to be 2 but it is unclear from the spec.  It does make sense that no transition can happen in the transfer of ownership and this should be made explicit in the spec.

**Comment by lentinem** [comment]
I agree that 1 should not work. 2 definitely should but it's unclear to me if 3 is allowed or not but either way I do agree the spec needs to be updated to reflect if 3 is supported or not.

**Comment by krOoze** [comment]
So I made an code sample exploring this: https://github.com/krOoze/Hello_Triangle/tree/queue_transfer

For the case 1: it works on my (AMD) platform, but does confuse the layers (they show errors)

For ~ case 3 (everything same but layout transition only in the releasing barrier): also works and layers do not complain.

I kinda don't agree with 2. There's no reason to pass this opportunity to transition the layout and outright ban it.

Okay, this seems to be followed in KhronosGroup/Vulkan-Docs#347 and KhronosGroup/Vulkan-Docs#346 .
With one corner case explored in KhronosGroup/Vulkan-Docs#344 .

**Comment by ParticlePeter** [comment]
Thanks for the clarification, I will observe the corresponding Vulkan-Docs issue(s). From my side this issue here can be closed.
 
@krOoze, thanks for your example.

**Comment by krOoze** [comment]
@ParticlePeter You can close your own Issues,
but I would wait to see resolution in the spec to judge the correctness of layers.

**Comment by mark-lunarg** [comment]
We're going to close for now, seems like the spec folks agreed that it was a spec but but I haven't seen any movement.

**Comment by ParticlePeter** [comment]
I haven't closed it due to krOoze's advice but I agree that its rather a spec issue and am following the related issues there. So from my side closing is fine.

**Comment by krOoze** [comment]
@ParticlePeter Nevermind me.
The sameness problem (KhronosGroup/Vulkan-Docs#347) was resolved. So probably layer changes will be needed (if they haven't been done already).

As for the remaining Issue, my bet is layout transitions will be allowed (because why not?). And that the meaning will be "one transition" despite two Barrier definitions (because of how the sameness is defined now).

**Comment by ParticlePeter** [comment]
@KrOoze, I don't see how it was resolved, v1.0.33 is still as unclear as v1.0.28 when one sentence was added which hardly clarifies anything (you pointed it out in still opened [KhronosGroup/Vulkan-Docs#347](https://github.com/KhronosGroup/Vulkan-Docs/issues/346). So why was Vulkan-Docs#347 closed? What do I miss?

**Comment by TobiasHector** [comment]
@ParticlePeter There's a whole bunch of weird interactions here, but the final actual fix should land in a week or two with the resolution of KhronosGroup/Vulkan-Docs#132 (which will also close KhronosGroup/Vulkan-Docs#346).

NB: Out of the three options presented above *1* (which apparently everyone thought was nonsense) is actually the correct one. Surprise! Actually I'm going to re-open this issue because of that, because I'm not sure what the heck the current validation layers do here. I'll also comment on KhronosGroup/Vulkan-Docs#346 to clarify the direction here. This issue probably shouldn't be closed until #346 gets a resolution, just in case.

**Comment by mark-lunarg** [comment]
@TobiasHector, validation is currently implementing option 2 so we will need to update for option 1 once the changes land in the spec.

**Comment by krOoze** [comment]
@mark-lunarg 
I have problem reproducing the error message again (using approach 1 in my example and 1.0.30 SDK), so there might be something else going on (unless messed up something).

**Comment by krOoze** [comment]
NVM apparently did mess up

**Comment by ParticlePeter** [comment]
Thanks for clarifying Tobias.

**Comment by ParticlePeter** [comment]
Added a comment to https://github.com/KhronosGroup/Vulkan-Docs/issues/346

**Comment by TobiasHector** [comment]
@ParticlePeter Does this issue not need to stay open so the validation layers can implement the correct thing now the spec has landed? Or would you rather a new issue be opened...? Re-opening for now, but you can close if you want a new issue opened rather than piggy-backing on yours :)

**Comment by ParticlePeter** [comment]
Well I assumed sure that the layers will follow soon with the spec clarification. But I must admit that I am not testing the layout transition while ownership transfer, I am transitioning in the render pass. Was afraid that I might miss the validation layer fix :-)

**Comment by mark-lunarg** [comment]
Lemme check on Monday, I don't want this one to get missed either.

**Comment by andens** [comment]
In addition to the error with layout transition, it should probably be mentioned that there seem to be errors concerning access masks as well. While the 1.0.39 spec clarifies the layout transition, it also recommends using 0 for destination access in the release barrier and 0 for source access in the acquire barrier. Using 0 in destination access for a release barrier gives a validation error about required access bit w.r.t. the layout.

**Comment by jzulauf-lunarg** [comment]
Should be fixed by 2521 above.
